sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -452,26 +496,67 @@ class SpillableDeviceBufferHandle private (
     materialized
   }
 
+  // used to gate when a spill is actively being done so that a second thread won't
+  // also begin spilling, and a handle won't release the underlying buffer if it's
+  // closed while spilling
+  private var toSpill: Option[DeviceMemoryBuffer] = None
+
+  // used to ensure the resource gets cleaned up properly after spilling and avoids
+  // double freeing
+  private var spilled: Option[DeviceMemoryBuffer] = None
+  override def releaseDeviceResource(): Unit = {
+    synchronized {
+      spilled.foreach(_.close())
+      spilled = None
+    }
+    super.releaseDeviceResource()
+  }
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      synchronized {
-        if (host.isEmpty && dev.isDefined) {
-          host = Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(dev.get))
-          sizeInBytes
+      val thisThreadSpills = synchronized {
+        if (host.isEmpty && dev.isDefined && toSpill.isEmpty) {
+          toSpill = dev
+          // incRefCount here so that if close() is called
+          // while we are spilling, we will prevent the buffer being freed
+          toSpill.get.incRefCount()
+          true
         } else {
-          0L
+          false
+        }
+      }
+      if (thisThreadSpills) {
+        withResource(toSpill.get) { buf =>
+          // the spill IO is non-blocking as it won't impact dev or host directly
+          // instead we "atomically" swap the buffers below once they are ready
+          val stagingHost =
+            Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(buf))
+          synchronized {
+            if (!closed) {
+              host = stagingHost
+            }
+            spilled = dev
+            dev = None
```

Comment:
I think we need to have code to handle the case where close was called in the middle of a spill.

It looks like we are going to leak stagingHost, and dev which is then stored into spilled with the current code.